File: app/src/main/java/com/doubean/ford/data/repository/NetworkBoundResource.java
package com.doubean.ford.data.repository;


import androidx.annotation.MainThread;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.WorkerThread;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MediatorLiveData;

import com.doubean.ford.util.AppExecutors;

import java.util.Objects;

/**
 * Simplified version of https://github.com/android/architecture-components-samples/blob/main/GithubBrowserSample/app/src/main/java/com/android/example/github/repository/NetworkBoundResource.kt
 * -No resource/response encapsulation
 * -Will only be updated once, no unnecessary loads
 * Original comment:
 * <p>
 * A generic class that can provide a resource backed by both the sqlite database and the network.
 * <p>
 * You can read more about it in the <a href="https://developer.android.com/arch">Architecture
 * Guide</a>.
 *
 * @param <ResultType>
 * @param <RequestType>
 */
public abstract class NetworkBoundResource<ResultType, RequestType> {
    private final AppExecutors appExecutors;

    private final MediatorLiveData<ResultType> result = new MediatorLiveData<>();

    /**
     * Note that the returned LiveData should always be exactly set once (by BumbleBee202111)
     */
    @MainThread
    NetworkBoundResource(AppExecutors appExecutors) {
        this.appExecutors = appExecutors;
        result.setValue(null);
        LiveData<ResultType> dbSource = loadFromDb();
        result.addSource(dbSource, data -> {
            result.removeSource(dbSource);
            if (shouldFetch(data)) {
                fetchFromNetwork(data);
            } else {
                result.addSource(dbSource, this::setValue);
            }
        });
    }

    @MainThread
    private void setValue(ResultType newValue) {
        if (!Objects.equals(result.getValue(), newValue)) {
            result.setValue(newValue);
        }
    }

    private void fetchFromNetwork(final ResultType cachedData) {
        LiveData<RequestType> apiResponse = createCall();
        result.addSource(apiResponse, response -> {
            result.removeSource(apiResponse);
            if (response != null) {
                appExecutors.diskIO().execute(() -> {
                    saveCallResult(response);
                    appExecutors.mainThread().execute(() ->
                            result.addSource(loadFromDb(),
                                    this::setValue)
                    );
                });
            } else {
                onFetchFailed();
                setValue(cachedData);
            }

        });
    }

    protected void onFetchFailed() {

    }

    public LiveData<ResultType> asLiveData() {
        return result;
    }

    @WorkerThread
    protected ResultType processResponse(ResultType response) {
        return response;
    }

    @WorkerThread
    protected abstract void saveCallResult(@NonNull RequestType item);

    @MainThread
    protected abstract boolean shouldFetch(@Nullable ResultType data);

    @NonNull
    @MainThread
    protected abstract LiveData<ResultType> loadFromDb();

    @NonNull
    @MainThread
    protected abstract LiveData<RequestType> createCall();
}

File: app/src/main/java/com/doubean/ford/data/repository/NetworkBoundResource.kt
/*
 * Copyright (C) 2017 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.doubean.ford.data.repository

import com.doubean.ford.api.ApiEmptyResponse
import com.doubean.ford.api.ApiErrorResponse
import com.doubean.ford.api.ApiResponse
import com.doubean.ford.api.ApiSuccessResponse
import com.doubean.ford.model.Resource
import kotlinx.coroutines.flow.*

/**
 * A generic class that can provide a resource backed by both the sqlite database and the network.
 *
 *
 * You can read more about it in the [Architecture
 * Guide](https://developer.android.com/arch).
 * @param <ResultType>
 * @param <RequestType>
</RequestType></ResultType> */

abstract class NetworkBoundResource<ResultType, RequestType> {

    private val result = flow {
        val dbSource = loadFromDb()
        val dbData = dbSource.first()
        emit(Resource.loading(dbData))
        if (shouldFetch(dbData)) {
            val networkSource = fetchFromNetwork(dbSource)
            emitAll(networkSource)
        } else {
            emitAll(dbSource.map { newData ->
                Resource.success(newData)
            })
        }
    }

    private suspend fun fetchFromNetwork(
        dbSource: Flow<ResultType?>,
    ) = when (val apiResponse = createCall()) {
        is ApiSuccessResponse -> {
            saveCallResult(apiResponse.body)
            dbSource.map { newData ->
                Resource.success(newData)

            }
        }
        is ApiEmptyResponse -> {
            dbSource.map { newData ->
                Resource.success(newData)
            }
        }
        is ApiErrorResponse -> {
            onFetchFailed()
            dbSource.map { newData ->
                Resource.error(
                    apiResponse.errorMessage,
                    newData
                )
            }
        }
    }

    protected open suspend fun onFetchFailed() {}

    fun asFlow() = result

    protected abstract suspend fun saveCallResult(item: RequestType)

    protected open fun shouldFetch(data: ResultType?) = true

    protected abstract fun loadFromDb(): Flow<ResultType?>

    protected abstract suspend fun createCall(): ApiResponse<RequestType>
}



